New world functions

samr46 opened the pull request:
Fixes #385 
Related Mantis issues at https://bugs.mtasa.com/view.php?id=8113 (see relations)

Added new property for [setWorldSpecialPropertyEnabled](https://wiki.multitheftauto.com/wiki/SetWorldSpecialPropertyEnabled):

"**watercreatures**" toggle randomly created water creatures (default: true) ([picture 1](https://i.imgur.com/v3PYmIF.jpg))

Added new function "**addBirds**" ([picture 2](https://i.imgur.com/g37GgfR.png)).

> Syntax: addBirds ( float startX, float startY, float startZ, float destX, float destY, float destZ, int count, int birdType, bool checkObstacles )

count: 1 - 6
birdType: 0 - 2

Returns the number of created birds if the function is successful, false otherwise.

Note: GTA:SA can render only 6 birds at the same time. You can use removeBirds before calling addBirds if you need to create 6 birds in one call.

Added new function "**removeBirds**" which removes all existing birds in game. It can be useful with **addBirds** function.

> Syntax: removeBirds ( )

Returns true if the function is successful, false otherwise.

Changed behavior of [setBirdsEnabled](https://wiki.multitheftauto.com/wiki/SetBirdsEnabled) to allow addBirds create new birds even if it's disabled.

---

Added new functions to control default **time cycle** (24 weathers, 8 cycles).

> Syntax: setTimeCycle ( string property, [float value = -1], [int weatherID = -1], [int cycle = -1] )

property:
- AmbientRed
- AmbientGreen
- AmbientBlue
- AmbientObjectRed
- AmbientObjectGreen
- AmbientObjectBlue
- SkyTopRed
- SkyTopGreen
- SkyTopBlue
- SkyBottomRed
- SkyBottomGreen
- SkyBottomBlue
- SunCoreRed
- SunCoreGreen
- SunCoreBlue
- SunCoronaRed
- SunCoronaGreen
- SunCoronaBlue
- SunSize
- SpriteSize
- SpriteBrightness
- ShadowStrength
- LightShadowStrength
- PoleShadowStrength
- FarClip
- FogStart
- LightsOnGroundBrightness
- LowCloudsRed
- LowCloudsGreen
- LowCloudsBlue
- BottomCloudsRed
- BottomCloudsGreen
- BottomCloudsBlue
- WaterRed
- WaterGreen
- WaterBlue
- WaterAlpha
- ColourCorrectRed
- ColourCorrectGreen
- ColourCorrectBlue
- ColourCorrectAlpha
- ColourAddRed
- ColourAddGreen
- ColourAddBlue
- ColourAddAlpha
- LowCloudsAlpha
- UpperCloudsAlpha
- Illumination

value (default: -1):
-1, false, nil - reset to original
0 - 255 for all properties except:
- SunSize (0.0 - 25.5)
- SpriteSize (0.0 - 25.5)
- SpriteBrightness (0.0 - 25.5)
- LightsOnGroundBrightness (0.0 - 25.5)
- FarClip (0 - 65535)
- FogStart (0 - 65535)

weatherID (default: -1): 0 - 23 (-1 set for all weathers)
cycle (default: -1): 0 - 7 (-1 set for all cycles)

Returns true if the function is successful, false otherwise.

> Syntax: getTimeCycle ( int weatherID, [int cycle = -1], [string property = ""] )
> Syntax: getOriginalTimeCycle ( int weatherID, [int cycle = -1], [string property = ""] )

weatherID: 0 - 23
cycle (default: -1): 0 - 7 (-1 retrieve property values of all cycles)
property (default: ""): the name of the property from setTimeCycle

Returns the number if cycle and property are specified or table if the function is successful, false otherwise.

> resetTimeCycle ( )

Reset all time cycle properties to original values.

Returns true if the function is successful, false otherwise.

samr46 committed:
setPlantsEnabled and setWaterCreaturesEnabled

New functions:
bool getPlantsEnabled()
bool setPlantsEnabled(bool enabled)
bool getWaterCreaturesEnabled()
bool setWaterCreaturesEnabled(bool enabled)

CrosRoad95 commented:
there is https://wiki.multitheftauto.com/wiki/SetWorldSpecialPropertyEnabled "randomfoliage", you could make option to place plant, that would be interesting.

samr46 commented:
Thanks. I was looking for this option but somehow never noticed "randomfoliage". My mistake.
"addPlant" function is an interesting idea.

Pirulax commented:
use setWorldSpecialPropertyEnabled for these things too.
That'd make sense.
Actually, there is no counterpart of setWorldSpecialPropertyEnabled, so you could make a isWorldSpecialPropertyEnabled function.

samr46 commented:
There is [isWorldSpecialPropertyEnabled](https://wiki.multitheftauto.com/wiki/IsWorldSpecialPropertyEnabled). I think it would be better if i add water creatures option in setWorldSpecialPropertyEnabled for now.

samr46 committed:
Removed plants functions & water creatures functions moved to setWorldSpecialPropertyEnabled

samr46 commented:
bool addBirds ( float startX, float startY, float startZ, float destX, float destY, float destZ, int count, int type, bool checkObstacles )

[Picture](https://i.imgur.com/g37GgfR.png)

But GTA:SA can render only 6 birds at the same time :(
What do you think?

CrosRoad95 commented:
six only ? 🤔 there could be possible to overwrite  condition `if birds <= 6 then create() end`

Pirulax commented:
I think @samr46 means that GTA has a hardcoded limit of 6 birds.

samr46 committed:
addBirds function
Syntax: int addBirds ( float startX, float startY, float startZ, float destX, float destY, float destZ, int count, int birdType, bool checkObstacles )

birdType (0 - 2)

Returns the number of created birds if the function is successful, false otherwise

Pirulax reviewed (changes requested):
You can create variables in one like, like so:
`int iBirdsCount, iBirdType`
You don't need to assign a value to them.

Pirulax commented:
lol, I'm too bastard to even review code 🗡 
Nah, you got it.

samr46 committed:
Code improvements

Einheit-101 commented:
is it possible to modify setWorldSpecialPropertyEnabled() in such a way that even currently streamed in foliage/birds get deleted? Currently you have to stream out/in to remove already existing elements.

samr46 commented:
Actually "randomfoliage" option instantly removes already existing foliage elements (grass, plants, seabed rocks). But if you enable foliage it will not create new elements around you until you stream out/in. By the way my implementation of foliage function did it instantly in both cases.

Also it shouldn't affect birds since it's not foliage. There is setBirdsEnabled() to disable birds generation.
I can add function to remove all existing birds (like a reset). Maybe it can be useful with addBirds function.

samr46 committed:
removeBirds, randombirds, code improvements

samr46 commented:
I added **removeBirds** function like suggested.

Also i added option "**randombirds**". It's needed to get a full control on birds in game (i described it in a pull request description).
I could edit [setBirdsEnabled](https://wiki.multitheftauto.com/wiki/SetBirdsEnabled) for such purpose but i think we should keep option to disable all flying birds (including customly created by addBirds).

samr46 committed:
Time cycle functions

samr46 commented:
I added the last planned feature for this pull request.

setTimeCycle function allows you to change different world properties like colors (note: sky gradients are smoother than setSkyGradient and it's cool), lights, shadows. Basically all GTA:SA timecyc options.
It only overrides default values so game continues to change world properties smoothly.
Also it can be easily synchronized between players (custom parser for timecyc.dat file on the client side is probably the most useful way).

More detailed info in the pull request description.

qaisjp commented:
This looks nice. I think the functionality of setBirdsEnabled should changed to perform `setWorldSpecialPropertyEnabled("randombirds", false)` followed by `removeBirds()` - I think addBirds should always work

samr46 committed:
setBirdsEnabled functionality & code improvements

samr46 committed:
Code improvements

qaisjp commented on Client/game_sa/CTimeCycleSA.h:
```diff
@@ -0,0 +1,135 @@
+ *
+ *****************************************************************************/
+
+#ifndef __CGAMESA_TIMECYCLE
```
Please use `#pragma once`

qaisjp committed:
Merge branch 'master' into new-world-functions

qaisjp committed:
Fix include guards

samr46 committed:
Predefined original values, code improvements

patrikjuvonen committed:
Merge remote-tracking branch 'upstream/master' into new-world-functions

patrikjuvonen commented on Client/mods/deathmatch/logic/lua/CLuaFunctionDefs.World.cpp:
```diff
@@ -1764,6 +1764,263 @@ int CLuaFunctionDefs::GetBirdsEnabled(lua_State* luaVM)
+    if (iCycleNum < -1 || iCycleNum > 7)
+        argStream.SetCustomError("The cycle should be between 0 and 7");
+
+    if (!argStream.HasErrors())
```
These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

patrikjuvonen commented on Client/mods/deathmatch/logic/lua/CLuaFunctionDefs.World.cpp:
```diff
@@ -1764,6 +1764,263 @@ int CLuaFunctionDefs::GetBirdsEnabled(lua_State* luaVM)
+    argStream.ReadNumber(iCycleNum, -1);
+    argStream.ReadString(strPropertyName, "");
+
+    if (!argStream.HasErrors())
```
These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

patrikjuvonen commented on Client/mods/deathmatch/logic/lua/CLuaFunctionDefs.World.cpp:
```diff
@@ -1764,6 +1764,263 @@ int CLuaFunctionDefs::GetBirdsEnabled(lua_State* luaVM)
+                if (fValue != -1)
+                    lua_pushnumber(luaVM, fValue);
+                else
+                    argStream.SetCustomError("Property doesn't exist");
```
I don't think this error is being outputted.

patrikjuvonen commented on Client/mods/deathmatch/logic/lua/CLuaFunctionDefs.World.cpp:
```diff
@@ -1764,6 +1764,263 @@ int CLuaFunctionDefs::GetBirdsEnabled(lua_State* luaVM)
+    argStream.ReadNumber(iCycleNum, -1);
+    argStream.ReadString(strPropertyName, "");
+
+    if (!argStream.HasErrors())
```
These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

patrikjuvonen commented on Client/mods/deathmatch/logic/lua/CLuaFunctionDefs.World.cpp:
```diff
@@ -1764,6 +1764,263 @@ int CLuaFunctionDefs::GetBirdsEnabled(lua_State* luaVM)
+                if (fValue != -1)
+                    lua_pushnumber(luaVM, fValue);
+                else
+                    argStream.SetCustomError("Property doesn't exist");
```
I don't think this error is being outputted.

patrikjuvonen commented on Client/mods/deathmatch/logic/lua/CLuaFunctionDefs.World.cpp:
```diff
@@ -1764,6 +1764,263 @@ int CLuaFunctionDefs::GetBirdsEnabled(lua_State* luaVM)
+    argStream.ReadNumber(iBirdType, 0);
+    argStream.ReadBool(bCheckObstacles, false);
+
+    if (!argStream.HasErrors())
```
These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

Woovie replied:
Is this an example of how to do it right?

Before
```cpp
    if (!argStream.HasErrors())
    {
        lua_pushboolean(luaVM, g_pGame->GetTimeCycle()->SetTimeCycle(strPropertyName, fValue, iWeatherID, iCycleNum));
        return 1;
    }
    else
        m_pScriptDebugging->LogCustom(luaVM, argStream.GetFullErrorMessage());

    lua_pushboolean(luaVM, false);
    return 1;
```

After
```cpp
    if (argStream.HasErrors())
    {
        return luaL_error(luaVM, argStream.GetFullErrorMessage());
    }
    else
        lua_pushboolean(luaVM, g_pGame->GetTimeCycle()->SetTimeCycle(strPropertyName, fValue, iWeatherID, iCycleNum));
        return 1;

    lua_pushboolean(luaVM, false);
    return 1;
```

qaisjp replied:
```cpp
    if (argStream.HasErrors())
    {
        return luaL_error(luaVM, argStream.GetFullErrorMessage());
    }

    lua_pushboolean(luaVM, g_pGame->GetTimeCycle()->SetTimeCycle(strPropertyName, fValue, iWeatherID, iCycleNum));
    return 1;
```

Is probably what you want. But you should probably use the new parser instead of the old parser.

Pirulax replied:
Yeah I suggest using the new parser as well:
```cpp
bool CLuaFunctionDefs::AddBirds(CVector start, CVector dest, std::optinal<unsigned int> birdsCount, std::optinal<unsigned char> birdType, std::optinal<bool> checkObstacles)
{
    // bool addBirds ( float startX, float startY, float startZ, float destX, float destY, float destZ, [int birdsCount = 1, int birdType = 0, bool checkObstacles = false] )
    return CStaticFunctionDefinitions::AddBirds(start, end, birdsCount.value_or(1), birdType.value_or(0), checkObstacles.value_or(false));
}
```

qaisjp committed:
Merge branch 'master' into new-world-functions

Woovie committed:
Merge branch 'master' of github.com:multitheftauto/mtasa-blue into new-world-functions

Pirulax commented on Client/game_sa/CGameSA.cpp:
```diff
@@ -693,6 +705,37 @@ bool CGameSA::IsUnderWorldWarpEnabled()
+    }
+    m_bWaterCreatures = bEnabled;
+}
+
```
```cpp
void CGameSA::SetWaterCreaturesEnabled(bool bEnabled)
{
    if (bEnabled == m_bWaterCreatures)
        return;

    using Func_t = int(__thiscall*)(void*); 
    (Func_t(bEnabled ? 0x6E3F90 : 0x6E3FD0))() // Init/Deinit WaterCreateManager

    m_bWaterCreatures = bEnabled;
}
```

TheNormalnij commented on Client/game_sa/CTimeCycleSA.h:
```diff
@@ -0,0 +1,132 @@
+#define VAR_CTimeCyc_UpperCloudsAlphaPool       0xB7A148
+#define VAR_CTimeCyc_IlluminationPool           0xB79FD8
+
+#define TIMECYCLE_AMBIENT_RED                   "AmbientRed"
```
I think we can use enum class for this, if we use new argument parser.

Commit:
Merge remote-tracking branch 'mta_main/master' into new-world-functions

Commit:
Move new functions to CLuaWorldDefs

Commit:
Used new arg parser for AddBirds and RemoveBirds functions

Commit:
New arg parser for ResetTimeCycle

Commit:
Apply compact format for SetWaterCreaturesEnabled

TheNormalnij commented:
I'm going to refactor this PR to using new argument parser and array for m_TimeCycleProps instead std::map. PR will be ready to merge after this changes.

christophercampbell1 commented:
Whats up with this, its a nice feature

Lpsd commented:
As per the previous messages in this thread, it was closed due to inactivity.

samr46 commented:
No point to revive this. I think I'll send new pull request with different (more useful) implementation of timecyc features when I get time to finish it. And separate pull requests for addBirds() or "watercreatures" features if someone needs it.
